### PR TITLE
Fix HDFS support for CU5+ BDC instances

### PR DESCRIPTION
**Pull request by Charles-Gagnon**
In CU5 a fix is being made so that the Knox gateway username is no longer going to always be root - instead it'll be whatever the initial username that was specified during deployment.

There's no direct way to query this username so instead we use the SQL login since that by default is going to be the same. This can be modified though (or a new login is made on the SQL instance after deployment) and so in that case we have to just do our best and prompt the user to enter in the username manually. For this first fix this means that user will be reprompted each time they reconnect to the server.

As part of this work I started the implementation of the BDC extension API so other extensions can use that to query BDC resources much more easily.

### CODE REVIEW

File: extensions/big-data-cluster/src/bigDataCluster/controller/apiGenerated.ts [hunk 1]
```diff
@@ -859,22 +859,12 @@ export class BdcRouterApi {
...
-        if (xRequestId === null || xRequestId === undefined) {
-            throw new Error('Required parameter xRequestId was null or undefined when calling getCluster.');
-        }
-
-        // verify required parameter 'connection' is not null or undefined
-        if (connection === null || connection === undefined) {
-            throw new Error('Required parameter connection was null or undefined when calling getCluster.');
-        }
-
         localVarHeaderParams['X-Request-Id'] = ObjectSerializer.serialize(xRequestId, "string");
```
Comment (chlafreniere) [inline extensions/big-data-cluster/src/bigDataCluster/controller/apiGenerated.ts:868]: what happens in objectserializer if xRequestId or connection is null/undefined?

File: extensions/mssql/src/sqlClusterLookUp.ts [hunk 2]
```diff
@@ -94,8 +96,16 @@ async function createSqlClusterConnInfo(sqlConnInfo: azdata.IConnectionProfile |
 	clusterConnInfo.options[constants.knoxPortPropName] = hostAndIp.port || constants.defaultKnoxPort;
 	let authType = clusterConnInfo.options[constants.authenticationTypePropName] = sqlConnInfo.options[constants.authenticationTypePropName];
 	if (authType && authType.toLowerCase() !== constants.integratedAuth) {
-		clusterConnInfo.options[constants.userPropName] = 'root'; //should be the same user as sql master
+		clusterConnInfo.options[constants.userPropName] = sqlConnInfo.options[constants.userPropName]; //should be the same user as sql master
 		clusterConnInfo.options[constants.passwordPropName] = credentials.password;
+		try {
+			const bdcApi = <bdc.IExtension>await vscode.extensions.getExtension(bdc.constants.extensionName).activate();
```
Comment (chlafreniere) [inline extensions/mssql/src/sqlClusterLookUp.ts:102]: remind me... what happens if we try to call activate multiple times?

Comment (Charles-Gagnon) [reply extensions/mssql/src/sqlClusterLookUp.ts:102]: Activate itself is only ever called once (unless the extension is deactivated). Future calls to activate will just return the API directly.

File: extensions/big-data-cluster/src/bigDataCluster/controller/apiGenerated.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/big-data-cluster/src/bigDataCluster/controller/apiGenerated.ts:868]: The header is empty. 

This is probably a result of a bug in the swagger spec where this was defined as being required but actually isn't (we fixed a number of these early on in the BDC development so this isn't a new thing). Calling it without these headers doesn't have any negative effect.

**Pull request merged**